added cron tutorial (#1515)
fermyon/developer: content/wasm-functions/using-cron-trigger.md

- [Prerequisites](#prerequisites)
- [Introduction to Cron Jobs](#introduction-to-cron-jobs)
  - [What Are Cron Jobs?](#what-are-cron-jobs)
- [Create Spin application](#create-spin-application)
  - [Deploy Spin Application to Fermyon Wasm Functions](#deploy-spin-application-to-fermyon-wasm-functions)
- [Creating a Cron Job for Your Spin Application](#creating-a-cron-job-for-your-spin-application)
- [Managing Cron Jobs](#managing-cron-jobs)
- [Delete Cron Job](#delete-cron-job)
- [Next Steps](#next-steps)
  

> This tutorial guides you through the process of scheduling HTTP requests in a Spin application running on Fermyon Wasm Functions using cron jobs. The `spin aka crons` command allows applications to receive scheduled HTTP triggers without relying on external services. In this tutorial, you'll learn how to define and manage cron jobs using the `spin aka crons` command.

Author: I also lean towards `cron` -- feels a bit more natural.

Also a good reminder I should flag that this feature is experimental in the tutorial

> This tutorial guides you through the process of scheduling HTTP requests in a Spin application running on Fermyon Wasm Functions using cron jobs. The `spin aka crons` command allows applications to receive scheduled HTTP triggers without relying on external services. In this tutorial, you'll learn how to define and manage cron jobs using the `spin aka crons` command.

Reviewer: This page seems to be largely the same content as the previous one but with the "cron trigger" language back in... is this meant to have been deleted in favour of the page above?  The TOC entry still points to this one.

* Log file rotation
* Scheduled notifications

# Create Spin application 

> You must have a running Spin application on Fermyon Wasm Functions to apply the cron trigger to. We'll create a new Spin application for scheduling tasks using cron jobs using the `http-js` template:

Author: You're right - it's not intended for scheduling tasks but the working is a bit misleading. I've reworded this sentence.

$ cd hello-cron-job

$ npm install
```

We'll go ahead add a specific path for our cron trigger in later steps:

```toml
[[trigger.http]]
route = "/hello"
component = "hello-cron-job"
```

Let's modify the response body so we can validate that our application is running on a specific interval. To do so, navigate to `src/index.js` and add the following snippet of code:

<!-- @selectiveCpy -->

```JavaScript
// For AutoRouter documentation refer to https://itty.dev/itty-router/routers/autorouter
import { AutoRouter } from 'itty-router';

let router = AutoRouter();

// Route ordering matters, the first route that matches will be used
// Any route that does not return will be treated as a middleware
// Any unmatched route will return a 404
router
    .get("/hello", () => {
    // Grab datetime
    const now = new Date().toISOString();
    // Return response with time of invocation
    return new Response(`Hello from a cron component at ${now}`, { status: 200 });
});

addEventListener('fetch', (event) => {
    event.respondWith(router.fetch(event.request));
});
```

To test locally, run the following command:

<!-- @selectiveCpy -->

```console
spin build --up
```

> If it is working as expected, you should see a message returned along the lines of (but with an accurate datetime):

Reviewer: To me this reads like I should see the message from `spin up`, but you actually mean when I curl it.

<!-- @selectiveCpy -->

```console
curl localhost:3000
```

<!-- @nocpy -->

```console

> Hello from a cron component at 2025-03-31T21:45:25.562Z

Reviewer: Having these as separate blocks kind of disconnects the response from the curl command.  Maybe rephrase along the lines of:

> Test the app to make sure it's working correctly by running:
> (curl block)
> You should see a response along the lines of:
> (response block)

```

## Deploy Spin Application to Fermyon Wasm Functions

> Before applying our cron trigger, we must have our Spin application running on the Fermyon Wasm Functions platform. Deploy the application with:

Reviewer: Reword cron trigger

<!-- @selectiveCpy -->

```console
spin aka deploy
```

Upon successful deployment, you should receive a domain name that you can use to test the application:

<!-- @nocpy -->

```console
curl https://ec8f3c23-b73d-4d51-aa69-aabd7ddc132a.aka.fermyon.tech/hello
Hello from a cron component at 2025-04-02T03:16:14.212Z% 
```

# Creating a Cron Job for Your Spin Application

> Now we will use the `spin aka crons` command to invoke the Spin app's HTTP endpoint at our desired interval. The syntax for this request follows standard crontab format.

Reviewer: ```suggestion
Now we will use the `spin aka crons` command to invoke the Spin app's HTTP endpoint on our desired schedule. The syntax for this request follows standard crontab format.
```

> Now we will use the `spin aka crons` command to invoke the Spin app's HTTP endpoint at our desired interval. The syntax for this request follows standard crontab format.

Reviewer: Link to the dialect of the crontab format that we support

> > `spin aka cron` supports single digits in each cron expression field and intervals in every position except the day of the week (e.g., `* */12 * * *` for every 12 hours). The smallest supported cron configuration is every 5 minutes (`*/5 * * * *`). Other syntax elements like comma-separated lists and ranges are not supported. Please feel free to[leave feedback](https://fibsu0jcu2g.typeform.com/to/G2u4tPcP) if your desired range is not supported. 

Reviewer: Is "single digits" correct?  You can't use double-digit numbers for e.g. day of month or hour of day?

> Let's create a cron job that runs every hour. Since scheduling is based on UTC, you may need to [convert your local time to UTC](https://www.worldtimebuddy.com/?pl=1&lid=100&h=100&hf=1) to ensure correct execution.

Reviewer: I'm not sure why local/UTC would matter for an "every hour" schedule?  I'm _guessing_ it's that the schedule is "every hour _on_ the hour" and you're thinking about half-hour timezone offsets?  Like okay I convert my local time to UTC, it's 5:08am or whatever, now what do I do with that result?

It may be better to pull out the "times are in UTC" comment to go with the scheduling/crontab discussion perhaps?

<!-- @selectiveCpy -->

```console

> spin aka crons create "0 * * * *" "/hello" "hello-cron-job"

Reviewer: Maybe explain what these fields are, since they are not named in the command.  I'm guessing the second is the route to be hit.  Is the third the name of the app to be hit?  Or the name of the job?  (In which case how does it know which app to hit?  Is this an "app in current directory" thing?)

If the third parameter is _not_ the name of the app, it would be good to choose a different string from the name of the app, so as to avoid ambiguity.
